crypto/crmf/crmf_asn.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2007-2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright Nokia 2007-2018
+ * Copyright Siemens AG 2015-2018
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ *
+ * CRMF implementation by Martin Peylo, Miikka Viljanen, and David von Oheimb.
+ */
+
+#include <openssl/asn1t.h>
+
+#include "crmf_int.h"
+
+/* explicit #includes not strictly needed since implied by the above: */
+#include <openssl/crmf.h>
+
+ASN1_SEQUENCE(OSSL_CRMF_PRIVATEKEYINFO) = {
+    ASN1_SIMPLE(OSSL_CRMF_PRIVATEKEYINFO, version, ASN1_INTEGER),
```

Comment:
Why don't you use the ASN1_EMBED & co for these not-optional fields ?
It can spare various allocations and indirections.
Or you wish to keep compatibility with pre-1.1.0 code ?

Reply:
The relevant code was written far before ASN1_EMBED became available. It would be beneficial for us not to use it yet, as we're aiming to maintain compatibility with 1.0.2 for the time being.